### PR TITLE
network_dialog: Validate hosts and ports

### LINKED ISSUE #2743: Poor error checking of socks proxy server address.
If the user configures EC network access via a Socks 5 server using a  numeric ip address that begins with a leading blank prior to the correct server ip address,, 
e.g " 192.168.8.229" instead of "192.168.8.229",
the user will not get any kind of error message, but no proxy connection will be made and the network status will remain disconnected, (red) if there is no server reachable except through a proxy. 

This bug is reproducible. Subsequent deletion of the leading blank character and closing the pop up window will allow the proxy to work.  Subsequent editing and adding the leading blank and closing the window will cause the proxy to fail, ...

This is a minor user experience problem.  This exists in EC 4.3.1,  and was also observed in Electrum 4.3.3.

### PR DESCRIPTION
**Pull request by EchterAgo**
This adds a new `HostValidator` class to validate hostnames and IP addresses entered into a `QLineEdit`. We use this class to validate the server and proxy host.

This also adds the existing `PortValidator` class to the server and proxy port widgets.

fixes #2743

### CODE REVIEW

**Comment by cculianu** [comment]
Tested. Looks good. I like the red on bad hostname, etc. Thanks for this.

Only issue is this: with this branch, if I click on the network dialog box, it takes like 10-20 seconds to materialize and in the meantime app hangs.

I switch to master, it no longer does this.

I initially thought maybe Python had to re-parse files and it would work ok if I quit app and came back. It still is slow. It hangs when I click on the network dialog basically... for a time the app is run. After the network dialog is constructed, hiding it and bringing it back is ok.

Something is funny here. See the below video for an idea of how long the delay is.

https://github.com/Electron-Cash/Electron-Cash/assets/266627/7cdd2e24-4902-4ccf-b265-2cd450a445f3

**Comment by cculianu** [comment]
Probably validating on DNS is slow A.F. and is the reason for the delay. Options include:

- set the validator on show event, not on c'tor? (might not really fix anything)
- do asynch DNS lookups?

File: electroncash_gui/qt/utils/validators.py [hunk 1]
```diff
@@ -54,15 +70,64 @@ def validate(self, inputStr: str, pos: int) -> QValidator.State:
...
+            ipaddress.ip_address(inputStr)
+            return set_state(QValidator.Acceptable)
+        except ValueError:
+            pass
+
+        # We try regex first to save a DNS lookup on clearly invalid names
+        if not self._domain_pattern.match(inputStr):
+            return set_state(QValidator.Invalid)
+
+        # Check for resolvable DNS
```
Comment (cculianu) [inline electroncash_gui/qt/utils/validators.py:120]: I suspect this is where the slow is.

Any chance we can just not do this?  I mean .. isn't it fine to just validate IP and something that *could* be a hostname and call it a day?

Comment (EchterAgo) [reply electroncash_gui/qt/utils/validators.py:120]: Yea, I'd be fine with that too. Will adjust soon.

Comment (EchterAgo) [reply electroncash_gui/qt/utils/validators.py:120]: I'll also adjust the regex to be a bit more correct.

Comment (EchterAgo) [reply electroncash_gui/qt/utils/validators.py]: Updated

**Comment by EchterAgo** [comment]
I just had a look at your video and I don't think the slowdown might not come from DNS, otherwise typing should be laggy as well and your general internet experience would not be a nice one.

**Comment by cculianu** [comment]
Who knows how DNS is implemented on crapOS.. maybe initial first calls are slower and then the lib decides on a server and it gets less slow. BTW typing was laggy in spurts for me.. not shown in video but it was noticeable at times.

**Comment by cculianu** [comment]
Hey so I tested it and it works great. Not as awesome as before in terms of validating it's a real host, sadly, which means it doesn't solve the problem as completely as you wanted.. I appreciate the sacrifice though so that at least it is fast and introduces no slowdown.  

I wouldn't have mentioned it but it made the app really annoying to use to have to wait 5-10s (sometimes 20s) to just get the network dialog up.  Made the app feel less great.. 

Code looks good.

Thanks. Approved and merging.

File: electroncash_gui/qt/utils/validators.py (hunk 1 above)
Comment (cculianu) [reply electroncash_gui/qt/utils/validators.py]: thanks man

**Pull request merged**